Make doctrine/annotations an optional dependency

W0rma opened the pull request:
Closes #333

goetas commented:
This looks nice, looking forward to it

goetas commented on README.md:
```diff
@@ -180,7 +191,7 @@ use JMS\Serializer\Annotation as Serializer;
 
 #[Serializer\XmlRoot('user')]
-#[Hateoas\Relation('self', href: "expr('/api/users/' ~ object.getId())")]
+#[Hateoas\Relation(name: 'self', href: "expr('/api/users/' ~ object.getId())")]
```
is the `name` argument needed? it seems that before, it was optional, making it look like annotations

W0rma replied:
Thank you for the hint.

It is not necessary anymore after 9e8a232bc00d9d97269712c33b67f088a87211e4

goetas commented on src/Representation/CollectionRepresentation.php:
```diff
@@ -7,15 +7,9 @@
 use JMS\Serializer\Annotation as Serializer;
 
-/**
- * @Serializer\ExclusionPolicy("all")
```
hmm., this change makes it mandatory to have the attribute driver, right? is this a BC break?

W0rma replied:
Yes, I think so.

I just added the annotations again to avoid that possible BC break.

goetas commented on src/Representation/OffsetRepresentation.php:
```diff
@@ -54,6 +54,13 @@
+#[Serializer\ExclusionPolicy('all')]
+#[Serializer\XmlRoot('collection')]
+#[Serializer\AccessorOrder(order: 'custom', custom: ['offset', 'limit', 'total'])]
+#[Hateoas\Relation(name: 'first', href: new Hateoas\Route(name: 'expr(object.getRoute())', parameters: 'expr(object.getParameters(0))', absolute: 'expr(object.isAbsolute())'))]
+#[Hateoas\Relation(name: 'last', href: new Hateoas\Route(name: 'expr(object.getRoute())', parameters: 'expr(object.getParameters((object.getTotal() - 1) - (object.getTotal() - 1) % object.getLimit()))', absolute: 'expr(object.isAbsolute())'), exclusion: new Hateoas\Exclusion(excludeIf: 'expr(object.getTotal() === null)'))]
+#[Hateoas\Relation(name: 'next', href: new Hateoas\Route(name: 'expr(object.getRoute())', parameters: 'expr(object.getParameters(object.getOffset() + object.getLimit()))', absolute: 'expr(object.isAbsolute())'), exclusion: new Hateoas\Exclusion(excludeIf: 'expr(object.getTotal() !== null && (object.getOffset() + object.getLimit()) >= object.getTotal())'))]
+#[Hateoas\Relation(name: 'previous', href: new Hateoas\Route(name: 'expr(object.getRoute())', parameters: 'expr(object.getParameters((object.getOffset() > object.getLimit()) ? object.getOffset() - object.getLimit() : 0))', absolute: 'expr(object.isAbsolute())'), exclusion: new Hateoas\Exclusion(excludeIf: 'expr(! object.getOffset())'))]
```
no problems at all in dropping older php versions 👍

goetas commented:
to me it looks good, i would just like to understand what happens is someone is using still annotations and uses some of the `Representation` classes?

W0rma commented:
> to me it looks good, i would just like to understand what happens is someone is using still annotations and uses some of the `Representation` classes?

@goetas Thank you for the review - I addressed all your comments.

After my latest adjustments the `Representation` classes are usable with or without annotations.

goetas commented:
@W0rma does your implementation allow to use both annotations and attributes at the same time? If yes, do we have a testcase in which both drivers are enabled?

W0rma commented:
> does your implementation allow to use both annotations and attributes at the same time? If yes, do we have a testcase in which both drivers are enabled?

@goetas Yes, it does. I added a test case which verifies that.

goetas commented:
lets ship it!

goetas commented:
tagged as https://github.com/willdurand/Hateoas/releases/tag/3.12.0-beta1
